### ISSUE TITLE
Fixing A recommender heavily biased towards popular items

### ISSUE THREAD

**Issue by MiladShahidi** [tensorflow/ranking#122]
(I know this is a StackExchange type question. But I haven't had much luck there with ranking-related questions).

I am training a pure CF recommender system on MovieLens 1M. I use embeddings to represent users and items and feed concatenated embeddings through two fully connected layers and a final single unit (sigmoid) to score the pair. The issue (below) happens whether I use approximate NDCG or `PAIRWISE_LOGISTIC_LOSS` with ndcg_lambda_weight (LambdaRank) as the loss function. (The graphs below use approximate NDCG).

The problem is that my model makes almost the exact same recommendations to everyone. When I fetch the top 5 scored items for 50 randomly selected users, all the 50 lists are identical and filled with famous and highly-rated movies.

I tried making the model smaller (smaller hidden layers and smaller embedding dimensions - 20 in graphs below) but these didn't help.

My interpretation is that it is safe (but not useful) to put those popular items at the top of the list, because you're likely to get some credit (e.g. NDCG@10) anyways. But it is not what I was expecting the model to learn.

Since it feels like I'm doing something wrong (overfitting?) I would like to diagnose this deeper before rushing to introduce diversity manually (like [this](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.459.8174&rep=rep1&type=pdf) perhaps).

Do you have any suggestions as to what might be going on and how to diagnose and avoid this?

These are the logits, losses and metric for a model with two layers (32 and 16 units), approximate ndcg loss, embedding dimension of 20. I tested the model around 40k and then 60k steps. The 40k model still had a little bit of variation in its recommendations. The 60k model was spitting out a constant list for everyone.

Logits: https://imgur.com/6QN0BKZ

Losses: https://imgur.com/VnInudE

NDCG@10: https://imgur.com/XKfNVls

**Comment by ramakumar1729** [comment]
@MiladShahidi : It is difficult to diagnose the exact problem. I would suggest the following: down-weighting for popular items in pairwise loss. If the clicked item is popular (assumed popularity is a curated list), multiply that loss by a small number. You can use `weight_feature_name` to generate and use such a feature.

**Comment by MiladShahidi** [comment]
Thank you very much @ramakumar1729. I really appreciate that.

I discovered today that I may be doing something wrong, but haven't pinned down the problem yet. When I replace all the MLP layers with a simple dot product between user and item embeddings (which essentially reduces it to a matrix factorizer optimized by SGD instead of ALS) the model manages to achieve a good MSE loss and a decent NDCG@10 and the recommendations are quite diverse across users.

However, when I change my scoring function to one that is almost identical to [this one](https://colab.research.google.com/github/tensorflow/ranking/blob/master/tensorflow_ranking/examples/handling_sparse_features.ipynb#scrollTo=B7Ft1i2oieEY&line=17&uniqifier=1) (without batch norm or dropout) it recommends the same thing to all users (even though it reaches a similar NDCG).

I'm trying to diagnose this and will post if I find something that might benefit the package or other users.

**Comment by ramakumar1729** [comment]
How are you measuring diversity? Can you share the numbers of diversity and NDCG for both approaches?

**Comment by ramakumar1729** [comment]
Closing this issue.
